src/Microsoft.OData.Core/UriParser/SemanticAst/ODataPath.cs

```diff
@@ -171,5 +193,38 @@ internal void AddRange(ODataPath oDataPath)
         {
             this.segments.AddRange(oDataPath.segments);
         }
+
+        public struct Enumerator : IEnumerator<ODataPathSegment>
+        {
+            private List<ODataPathSegment>.Enumerator enumerator;
+
+            internal Enumerator(List<ODataPathSegment> list)
+            {
+                this.enumerator = list.GetEnumerator();
+            }
+
+            public ODataPathSegment Current => this.enumerator.Current;
+
+            object IEnumerator.Current => this.enumerator.Current;
+
+            public void Dispose()
+            {
+                this.enumerator.Dispose();
+            }
+
+            public bool MoveNext()
+            {
+                return this.enumerator.MoveNext();
+            }
+
+            public void Reset()
+            {
+                // Unfortunately "Reset" is implemented explicitly and so we need to box IEnumerable<ODataPathSegment>, call "Reset",

```

Comment:
Is there a reason that we don't want to explicitly implement this as well? Honestly, I think I would rather just throw `NotSupportedException`

Reply:
That opens a potential regression, because `Reset()` would have worked prior to my refactoring.

Reply:
I'm reverting my enumeration change - the indexer gets me what I need perf wise

Reply:
> That opens a potential regression, because Reset() would have worked prior to my refactoring.

Great point, sorry about that